src/providers/twitch/PubSubClient.cpp

```diff
@@ -0,0 +1,214 @@
+#include "providers/twitch/PubSubClient.hpp"
+
+#include "common/QLogging.hpp"
+#include "providers/twitch/PubSubActions.hpp"
+#include "providers/twitch/PubSubHelpers.hpp"
+#include "providers/twitch/PubSubMessages.hpp"
+#include "providers/twitch/pubsubmessages/Unlisten.hpp"
+#include "singletons/Settings.hpp"
+#include "util/DebugCount.hpp"
+#include "util/Helpers.hpp"
+#include "util/RapidjsonHelpers.hpp"
+
+#include <exception>
+#include <thread>
+
+namespace chatterino {
+
+static const char *PING_PAYLOAD = "{\"type\":\"PING\"}";
+
+PubSubClient::PubSubClient(WebsocketClient &websocketClient,
+                           WebsocketHandle handle,
+                           const PubSubClientOptions &clientOptions)
+    : websocketClient_(websocketClient)
+    , handle_(handle)
+    , clientOptions_(clientOptions)
+{
+}
+
+void PubSubClient::start()
+{
+    assert(!this->started_);
+
+    this->started_ = true;
+
+    this->ping();
+}
+
+void PubSubClient::stop()
+{
+    assert(this->started_);
+
+    this->started_ = false;
+}
+
+void PubSubClient::close(const std::string &reason,
+                         websocketpp::close::status::value code)
+{
+    WebsocketErrorCode ec;
+
+    auto conn = this->websocketClient_.get_con_from_hdl(this->handle_, ec);
+    if (ec)
+    {
+        qCDebug(chatterinoPubSub)
+            << "Client::close(): Error getting con:" << ec.message().c_str();
+        return;
+    }
+
+    conn->close(code, reason, ec);
+    if (ec)
+    {
+        qCDebug(chatterinoPubSub)
+            << "Client::close(): Error closing:" << ec.message().c_str();
```

Comment:
minor nitpick: "hardcoding log message origins" is something that should be avoided since we make use of logging categories, which with `QT_MESSAGE_PATTERN` environment variable can show exact origin of the log message.
Example content of mentioned variable which is helpful: `"[%{time hh:mm:ss.zzz}] %{function} %{message}"`

Though by the way, I think that's something that should rather be specified in contribution guidelines to maintain specific and consistent 'format' for logging messages.